crates/biome_js_analyze/src/lint/nursery/no_adjacent_overload_signatures.rs

```diff
@@ -0,0 +1,332 @@
+
+declare_rule! {
+    /// Disallow non adjacent overload signatures.
+    ///
+    /// Overload signatures must be adjacent.
+    /// If a key is defined multiple times, only the last definition takes effect. Previous definitions are ignored.
+    /// This rule is useful for preventing accidental overloads that are not adjacent.
+    /// It is recommended to keep the overload signatures adjacent to make the code easier to read and maintain.
+    ///
+    /// Source: https://typescript-eslint.io/rules/adjacent-overload-signatures/
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    ///
+    /// type Foo = {
+    ///   foo_type(s: string): void;
+    ///   foo_type(n: number): void;
+    ///   bar_type(): void;
+    ///   foo_type(sn: string | number): void;
+    /// };
+    ///
+    /// interface Foo {
+    ///   foo_interface(s: string): void;
+    ///   foo_interface(n: number): void;
+    ///   bar_interface(): void;
+    ///   foo_interface(sn: string | number): void;
+    /// }
+    ///
+    /// class A {
+    ///   fooA(s: string): void;
+    ///   fooA(n: number): void;
+    ///   barA(): void {};
+    ///   fooA(sn: string | number): void {};
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// declare namespace Foo {
+    ///   export function foo_declare(s: string): void;
+    ///   export function foo_declare(n: number): void;
+    ///   export function foo_declare(sn: string | number): void;
+    ///   export function bar_declare(): void;
+    /// }
+    ///
+    /// type Foo = {
+    ///   foo_type(s: string): void;
+    ///   foo_type(n: number): void;
+    ///   foo_type(sn: string | number): void;
+    ///   bar_type(): void;
+    /// };
+    ///
+    /// interface Foo {
+    ///   foo_interface(s: string): void;
+    ///   foo_interface(n: number): void;
+    ///   foo_interface(sn: string | number): void;
+    ///   bar_interface(): void;
+    /// }
+    ///
+    /// class A {
+    ///   fooA(s: string): void;
+    ///   fooA(n: number): void;
+    ///   fooA(sn: string | number): void {}
+    ///   barA(): void {}
+    /// }
+    /// ```
+    ///
+    pub NoAdjacentOverloadSignatures {
+        version: "next",
+        name: "noAdjacentOverloadSignatures",
+        sources: &[
+            RuleSource::EslintTypeScript("adjacent-overload-signatures")
+        ],
+        recommended: false,
+    }
+}
+
+impl Rule for NoAdjacentOverloadSignatures {
+    type Query = Ast<JsModuleItemList>;
+    type State = Vec<(String, TextRange)>;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let mut methods: Vec<Vec<(SyntaxNodeText, usize, TextRange)>> = Vec::new();
+        let mut index = 0;
+        let mut export_vec = vec![];
+
+        for item in node {
+            if let AnyJsModuleItem::AnyJsStatement(node) = item {
```

Comment:
We should reduce code branching. In this case, use `let/else`:

```rs
let AnyJsModuleItem::AnyJsStatement(node) = item else {
	continue
};
```

With this, you can remove this `if`